### PR TITLE
GHI-486: Switch over to using LLVM utf8-checking code to better validate characters

### PR DESCRIPTION
**Pull request by timwoj**
This takes out the hand-validation of UTF-8 characters. Instead it uses the LLVM code we have checked in to find sequences in a string and either escape the bytes or insert them as characters. It fixes a few corner cases where bytes were being considered valid when they shouldn't have been. The unit test is updated with the corner cases as well.

### CODE REVIEW

File: src/util.cc [hunk 1]
```diff
@@ -1910,50 +1912,23 @@ string json_escape_utf8(const string& val)
...
-		
-		// if we didn't make it to the end of the character without finding an error, insert just this
-		// character and skip ahead. Otherwise insert all of the bytes for this character into the result.
-		if ( c_idx != idx + char_size )
+		// Find out how long the next character should be.
+		unsigned int char_size = getNumBytesForUTF8(val[idx]);
+
+		// If it says that it's a single character or it's not an invalid string UTF8 sequence, insert the one
+		// escaped byte into the string, step forward one, and go to the next character.
+		if ( char_size == 0 || isLegalUTF8Sequence(val_data+idx, val_data+idx+char_size) == 0 )
```
Comment (0xxon) [inline src/util.cc]: Just FYI - you forgot to check if there is enough data left in the string here. If you end a string with the beginning of a multi-byte UTF-8 sequence, isLegalUTF8Sequence will happily read over the string boundary.

Valgrind example output:
```
==17277== Thread 8 bro: ssh/Log::WR:
==17277== Conditional jump or move depends on uninitialised value(s)
==17277==    at 0xAEEF58: isLegalUTF8 (ConvertUTF.c:408)
==17277==    by 0xAEF0C7: isLegalUTF8Sequence (ConvertUTF.c:438)
==17277==    by 0xACDDC8: json_escape_utf8(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (util.cc:1920)
==17277==    by 0xC33AA5: threading::formatter::JSON::BuildJSON(threading::Value*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const (JSON.cc:156)
==17277==    by 0xC33178: threading::formatter::JSON::Describe(ODesc*, int, threading::Field const* const*, threading::Value**) const (JSON.cc:36)
==17277==    by 0xEDC6E4: logging::writer::Ascii::DoWrite(int, threading::Field const* const*, threading::Value**) (Ascii.cc:366)
==17277==    by 0xF012FB: logging::WriterBackend::Write(int, int, threading::Value***) (WriterBackend.cc:241)
==17277==    by 0xF033E7: logging::WriteMessage::Process() (WriterFrontend.cc:62)
==17277==    by 0xC298C3: threading::MsgThread::Run() (MsgThread.cc:379)
==17277==    by 0xC24BBB: threading::BasicThread::launcher(void*) (BasicThread.cc:208)
==17277==    by 0xC25685: void* std::__invoke_impl<void*, void* (*)(void*), threading::BasicThread*>(std::__invoke_other, void* (*&&)(void*), threading::BasicThread*&&) (invoke.h:60)
==17277==    by 0xC255FA: std::__invoke_result<void* (*)(void*), threading::BasicThread*>::type std::__invoke<void* (*)(void*), threading::BasicThread*>(void* (*&&)(void*), threading::BasicThread*&&) (invoke.h:95)
```

I am going to fix that during the merge - so there is nothing for you to do :)

**Comment by 0xxon** [comment]
This looks good besides the read-over-end-of-string-issue I pointed out above; I will fix that during the merge.

File: src/util.cc (hunk 1 above)
Comment (dwdixon) [reply src/util.cc]: @0xxon Thanks for this, I'm wondering if what you've spotted and fixed here when you merged could have been contributing to abnormally high memory utilization/operations/odd memory behavior?  I'm going to switch back to master now that this has been merged and your additional improvement included and see how things are looking before I report any other errors that may not have been related.  But I will open a few issues today/tomorrow that I am seeing in my reporter.log that probably need to be addressed.

Comment (jsiwek) [reply src/util.cc]: @dwdixon there was also a memory leak introduced in the `master` branch recently (last week or two).  A fix is pending here: https://github.com/zeek/zeek/pull/503, so might follow that and/or see if it helps resolve any new memory usage issues you see in `master`

Comment (dwdixon) [reply src/util.cc]: @jsiwek Ah, awesome, that very well could be what I've been seeing Jon I was thinking memory leak the whole time based on what I was observing at least at a high level but wasn't certain, thanks I will check out #503 !
